packages/@aws-cdk/aws-lambda-event-sources/lib/stream.ts

```diff
@@ -21,6 +21,50 @@ export interface StreamEventSourceProps {
    */
   readonly batchSize?: number;
 
+  /**
+   * If the function returns an error, split the batch in two and retry.
+   *
+   * @default false
+   */
+  readonly bisectBatchOnFunctionError?: boolean;
+
+  /**
+   * An Amazon SQS queue or Amazon SNS topic destination for discarded records.
+   *
+   * @default discarded records are ignored
+   */
+  readonly destinationOnFailure?: lambda.IDestination;
```

Comment:
Thinking about this more - `IDestination` here is not going to work. `IDestination` comes with a lot of implicit assumptions on where this is going to be bound. 

For instance, a user could very well accidentally wire up a [`LambdaDestination` construct](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda-destinations/lib/lambda.ts) makes assumptions around the format of the message that the destination receives and does some parsing around this. There's no indication anywhere that this is going to be the format of the message received in the case of event source failure destinations. Even if we programmatically restrict that only `SnsDestination` and `SqsDestination` can be used here, we'll be painting ourselves into a corner where these constructs can no longer make similar assumptions.

All we need here is an SQS or an SNS ARN. I suggest we model this as a new in this package called `IEventSourceDLQ` and extend two classes `SqsDLQ` and `SnsDLQ` with a `bind` method that returns the ARN. The pattern is still similar to the one used in `IDestination` but we're not reusing it. 

Also, I don't think there is a need to go with a separate package on this.

Reply:
> Also, I don't think there is a need to go with a separate package on this.

cc @eladb - what do you think of this?

Reply:
I had a separate conversation with @eladb regarding this and he agrees with my assessment. We need a separate set of destination interface and classes but it can be part of the lambda-event-sources package.

Reply:
Hoping to wrap up the integration test today.  I may need an assist in double checking it as I had to add a lot of things to overcome our security scanning.  I plan on migrating away from IDestination now that I've got a working integration test and hope to have something ready today or tomorrow.

Reply:
I ended up having to pull the new DLQ classes up to aws-lambda to prevent a circular dependency with aws-lambda-event-sources and aws-lambda.